src/main/java/org/tugraz/sysds/runtime/instructions/cp/TensorGenCPInstruction.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2019 Graz University of Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.tugraz.sysds.runtime.instructions.cp;
+
+import org.tugraz.sysds.common.Types;
+import org.tugraz.sysds.lops.TensorGen;
+import org.tugraz.sysds.runtime.DMLRuntimeException;
+import org.tugraz.sysds.runtime.controlprogram.context.ExecutionContext;
+import org.tugraz.sysds.runtime.data.TensorBlock;
+import org.tugraz.sysds.runtime.instructions.InstructionUtils;
+import org.tugraz.sysds.runtime.matrix.data.MatrixBlock;
+import org.tugraz.sysds.runtime.matrix.operators.Operator;
+import org.tugraz.sysds.runtime.util.UtilFunctions;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.StringTokenizer;
+
+public class TensorGenCPInstruction extends UnaryCPInstruction {
+	public static final String DELIM = " ";
+
+	private final CPOperand _opDims;
+	private final CPOperand _opByRow;
+
+	private TensorGenCPInstruction(Operator op, CPOperand in1, CPOperand in2, CPOperand in3, CPOperand out,
+	                               String opcode, String istr) {
+		super(CPType.TensorGen, op, in1, out, opcode, istr);
+		_opDims = in2;
+		_opByRow = in3;
+	}
+
+	public static TensorGenCPInstruction parseInstruction(String str) {
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		InstructionUtils.checkNumFields(parts, 4);
+		String opcode = parts[0];
+		CPOperand in1 = new CPOperand(parts[1]);
+		CPOperand in2 = new CPOperand(parts[2]);
+		CPOperand in3 = new CPOperand(parts[3]);
+		CPOperand out = new CPOperand(parts[4]);
+		if(!opcode.equalsIgnoreCase(TensorGen.TENSOR_OPCODE))
+			throw new DMLRuntimeException("Unknown opcode while parsing an TensorGenInstruction: " + str);
+		else
+			return new TensorGenCPInstruction(new Operator(true), in1, in2, in3, out, opcode, str);
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) {
+		TensorBlock out;
+		// TODO use byRow, or remove it
+		BooleanObject byRow = (BooleanObject) ec.getScalarInput(_opByRow.getName(), Types.ValueType.BOOLEAN,
+							_opByRow.isLiteral());
```

Comment:
Please don't spend too much time on this random data generator. Instead please call the matrix random data generator and reshape it to a tensor for now. The reason is that we need to implement this data generator with awareness of the blocking of distributed tensors in order to ensure that rand with specified seed yields exactly the same result in both local and distributed operations (with blocked output representations).

Reply:
I am actually not using a random data generator, I think I just misnamed the instruction. `TensorGen` is responsible to construct the tensor from some kind of given data, which could be a string, matrix, tensor or single scalar. We can discuss a new name in today's meeting.